File: packages/expo-updates/ios/EXUpdates/AppLauncher/EXUpdatesAppLauncherNoDatabase.m
//  Copyright © 2019 650 Industries. All rights reserved.

#import <EXUpdates/EXUpdatesAsset.h>
#import <EXUpdates/EXUpdatesAppLauncherNoDatabase.h>
#import <EXUpdates/EXUpdatesEmbeddedAppLoader.h>
#import <EXUpdates/EXUpdatesUtils.h>

NS_ASSUME_NONNULL_BEGIN

static NSString * const EXUpdatesErrorLogFile = @"expo-error.log";

@interface EXUpdatesAppLauncherNoDatabase ()

@property (nullable, nonatomic, strong, readwrite) EXUpdatesUpdate *launchedUpdate;
@property (nullable, nonatomic, strong, readwrite) NSURL *launchAssetUrl;
@property (nullable, nonatomic, strong, readwrite) NSMutableDictionary *assetFilesMap;

@end

/**
 * Implementation of EXUpdatesAppLauncher which always uses the update embedded in the application
 * package, avoiding SQLite and the expo-updates file store entirely.
 *
 * This is only used in rare cases when the database/file system is corrupt or otherwise
 * inaccessible, but we still want to avoid crashing. The exported property `isEmergencyLaunch` on
 * EXUpdatesModule should be `true` whenever this class is used.
 */
@implementation EXUpdatesAppLauncherNoDatabase

- (void)launchUpdateWithConfig:(EXUpdatesConfig *)config
{
  _launchedUpdate = [EXUpdatesEmbeddedAppLoader embeddedManifestWithConfig:config database:nil];
  if (_launchedUpdate) {
    if (_launchedUpdate.status == EXUpdatesUpdateStatusEmbedded) {
      NSAssert(_assetFilesMap == nil, @"assetFilesMap should be null for embedded updates");
      _launchAssetUrl = [[NSBundle mainBundle] URLForResource:EXUpdatesBareEmbeddedBundleFilename withExtension:EXUpdatesBareEmbeddedBundleFileType];
    } else {
      _launchAssetUrl = [[NSBundle mainBundle] URLForResource:EXUpdatesEmbeddedBundleFilename withExtension:EXUpdatesEmbeddedBundleFileType];

      NSMutableDictionary *assetFilesMap = [NSMutableDictionary new];
      for (EXUpdatesAsset *asset in _launchedUpdate.assets) {
        NSURL *localUrl = [EXUpdatesUtils urlForBundledAsset:asset];
        if (localUrl && asset.key) {
          assetFilesMap[asset.key] = localUrl.absoluteString;
        }
      }
      _assetFilesMap = assetFilesMap;
    }
  }
}

- (BOOL)isUsingEmbeddedAssets
{
  return _assetFilesMap == nil;
}

@end

NS_ASSUME_NONNULL_END

File: packages/expo-updates/ios/EXUpdates/AppLauncher/EXUpdatesAppLauncherNoDatabase.swift
//  Copyright © 2019 650 Industries. All rights reserved.

// swiftlint:disable force_unwrapping

import Foundation

/**
 * Implementation of EXUpdatesAppLauncher which always uses the update embedded in the application
 * package, avoiding SQLite and the expo-updates file store entirely.
 *
 * This is only used in rare cases when the database/file system is corrupt or otherwise
 * inaccessible, but we still want to avoid crashing. The exported property `isEmergencyLaunch` on
 * EXUpdatesModule should be `true` whenever this class is used.
 */
@objcMembers
public final class EXUpdatesAppLauncherNoDatabase: NSObject, EXUpdatesAppLauncher {
  public var launchedUpdate: EXUpdatesUpdate?
  public var launchAssetUrl: URL?
  public var assetFilesMap: [String: Any]?

  public override init() {}

  public func launchUpdate(withConfig config: EXUpdatesConfig) {
    launchedUpdate = EXUpdatesEmbeddedAppLoader.embeddedManifest(withConfig: config, database: nil)
    if let launchedUpdate = launchedUpdate {
      if launchedUpdate.status == EXUpdatesUpdateStatus.StatusEmbedded {
        precondition(assetFilesMap == nil, "assetFilesMap should be null for embedded updates")
        launchAssetUrl = Bundle.main.url(
          forResource: EXUpdatesEmbeddedAppLoader.EXUpdatesBareEmbeddedBundleFilename,
          withExtension: EXUpdatesEmbeddedAppLoader.EXUpdatesBareEmbeddedBundleFileType
        )
      } else {
        launchAssetUrl = Bundle.main.url(
          forResource: EXUpdatesEmbeddedAppLoader.EXUpdatesEmbeddedBundleFilename,
          withExtension: EXUpdatesEmbeddedAppLoader.EXUpdatesEmbeddedBundleFileType
        )

        var assetFilesMapLocal: [String: String] = [:]
        for asset in launchedUpdate.assets()! {
          if let assetKey = asset.key, let localUrl = EXUpdatesUtils.url(forBundledAsset: asset) {
            assetFilesMapLocal[assetKey] = localUrl.absoluteString
          }
        }
        assetFilesMap = assetFilesMapLocal
      }
    }
  }

  public func isUsingEmbeddedAssets() -> Bool {
    return assetFilesMap == nil
  }
}
